### ISSUE TITLE
startOf('hour') and endOf('hour') not DST aware

### ISSUE THREAD

**Issue by joelmukuthu** [moment/moment#2749]
For the hour when DST ended in the `ECT (Europe/Copenhagen)` timezone, `startOf('hour')` mutates the moment to the start of the previous hour instead of the current hour:

``` js
moment('2015-10-25T02:00:00.000+02:00').unix(); // 1445731200 (an hour before DST ended)
moment('2015-10-25T02:00:00.000+01:00').unix(); // 1445734800 (the extra hour)
moment('2015-10-25T03:00:00.000+01:00').unix(); // 1445738400 (the hour after DST ended)

// however, with startOf('hour'), this happens
moment('2015-10-25T02:00:00.000+02:00').startOf('hour').unix(); // 1445731200 (good)
moment('2015-10-25T02:00:00.000+01:00').startOf('hour').unix(); // 1445731200 (bad, this should be 1445734800)
moment('2015-10-25T03:00:00.000+01:00').startOf('hour').unix(); // 1445738400 (good)
```

Might be related to https://github.com/moment/moment/issues/1990. Tested on version 2.10.6

**Comment by joelmukuthu** [comment]
As discovered in https://github.com/moment/moment/issues/1990, this is also a problem in `Date.prototype.set{XXX}`:
```js
var date = new Date('2015-10-25T02:00:00+01:00');
date.toString(); // Sun Oct 25 2015 02:00:00 GMT+0100 (CET)
date.setMinutes(0);
date.toString(); // Sun Oct 25 2015 02:00:00 GMT+0200 (CEST)
```
The above happens in Chrome. In Firefox, it happens for the hour before DST ends:
```js
var date = new Date('2015-10-25T02:00:00+02:00');
date.toString(); // Sun Oct 25 2015 02:00:00 GMT+0200 (CEST)
date.setMinutes(0);
date.toString(); // Sun Oct 25 2015 02:00:00 GMT+0100 (CEST)
```
Both these cases can be avoided by using `setTime` instead of `setMinutes`, `setSeconds`... as described here http://stackoverflow.com/questions/9820468/adding-30-minutes-to-date-causes-it-to-go-back-30-minutes:
```js
var date = new Date('2015-10-25T02:00:00+01:00');
date.toString(); // Sun Oct 25 2015 02:00:00 GMT+0100 (CET)
date.setTime(0); // for startOf('hour')
date.setTime(date.getTime() + (59 * 60000) + (59 * 1000) + 999); // for endOf('hour')
date.toString(); // Sun Oct 25 2015 02:59:59 GMT+0100 (CET)
```
Could you adopt this implementation for moment?

**Comment by icambron** [comment]
Good analysis. I think that would be a reasonable fix, yeah.

**Comment by darakeon** [comment]
@joelmukuthu , maybe it's fixed, `endOf` for `day` seems fixed.

**Comment by joelmukuthu** [comment]
@darakeon doesn't seem fixed yet, I just tried with v2.20.1. I think it will be fixed by this PR though: https://github.com/moment/moment/pull/4338
